integration/instance.go

```diff
@@ -0,0 +1,392 @@
+	}
+
+	cmdName, err := BinaryPath()
+	if err != nil {
+		return err
+	}
+	cmdArgs := []string{
+		"run", // mode
+		fmt.Sprintf("--hostname=%s", obj.Hostname),
+		fmt.Sprintf("--client-urls=%s", obj.clientURL),
+		fmt.Sprintf("--server-urls=%s", obj.serverURL),
+		fmt.Sprintf("--prefix=%s", obj.tmpPrefixDirectory),
+		fmt.Sprintf("--converged-timeout=%d", convergedTimeout),
+		"--converged-timeout-no-exit",
+		fmt.Sprintf("--converged-status-file=%s", obj.convergedStatusFile),
+	}
+	if len(seeds) > 0 {
+		urls := []string{}
+		for _, instance := range seeds {
+			if instance.cmd == nil {
+				return fmt.Errorf("instance `%s` has not started yet", instance.Hostname)
+			}
+			urls = append(urls, instance.clientURL)
+		}
+		s := fmt.Sprintf("--seeds=%s", urls[0])
+		// TODO: we could just add all the seeds instead...
+		//s := fmt.Sprintf("--seeds=%s", strings.Join(urls, ","))
+		cmdArgs = append(cmdArgs, s)
+	}
+	obj.cmd = exec.Command(cmdName, cmdArgs...)
+	obj.cmd.Env = []string{
+		fmt.Sprintf("MGMT_TEST_ROOT=%s", obj.testRootDirectory),
+	}
+
+	// output file for storing logs
+	file, err := os.Create(path.Join(obj.dir, StdoutStderrFile))
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating log file")
+	}
+	defer file.Close()
+	obj.cmd.Stdout = file
+	obj.cmd.Stderr = file
+
+	if err := obj.cmd.Start(); err != nil {
+		return errwrap.Wrapf(err, "error starting mgmt")
+	}
+
+	return nil
+}
+
+// Kill the process immediately. This is a `kill -9` for if things get stuck.
+func (obj *Instance) Kill() error {
+	if obj.cmd == nil {
+		return nil // already dead
+	}
+	if obj.cmd.Process == nil {
+		return nil // nothing running
+	}
+
+	// cause a stack dump first if we can
+	_ = obj.cmd.Process.Signal(syscall.SIGQUIT)
+
+	return obj.cmd.Process.Kill()
+}
+
+// Quit sends a friendly shutdown request to the process. You can specify a
+// context if you'd like to exit earlier. If you trigger an early exit with the
+// context, then this will end up running a `kill -9` so it can return.
+func (obj *Instance) Quit(ctx context.Context) error {
+	if obj.cmd == nil {
+		return fmt.Errorf("no process is running")
+	}
+	if err := obj.cmd.Process.Signal(os.Interrupt); err != nil {
+		return errwrap.Wrapf(err, "could not send interrupt signal")
+	}
+
+	var err error
+	wg := &sync.WaitGroup{}
+	done := make(chan error)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		done <- obj.cmd.Wait()
+		close(done)
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case err = <-done:
+		case <-ctx.Done():
+			obj.Kill() // should cause the Wait() to exit
+			err = ctx.Err()
+		}
+	}()
+
+	wg.Wait()
+	obj.cmd = nil
+	return err
+}
+
+// Wait until the first converged state we hit. It is not necessary to use the
+// `--converged-timeout` option with mgmt for this to work. It tracks this via
+// the `--converged-status-file` option which can be used to track the varying
+// convergence status.
+func (obj *Instance) Wait(ctx context.Context) error {
```

Comment:
how does this handle race conditions? where mgmt instance/cluster is (partly) converged but Wait is not called?

Reply:
What do you mean?

Reply:
You need to set the converged-timeout to a high value, making waiting for it kinda useless as a sleep would almost suffice as well. (a low timeout is preferred to have faster tests). But with a to low value you have the risk for a configuration to converge after you deploy but before the moment you start waiting for convergence. 

Ideally you would wait for a specific deploy ID to converge instead of just any sign of convergence.